src/firefly/js/core/ComponentCntlr.js

```diff
@@ -0,0 +1,177 @@
+    flux.process({type: HIDE_ALL_DIALOGS, payload: {}});
+}
+
+export function dispatchComponentStateChange(componentId, componentState) {
+    flux.process({type: COMPONENT_STATE_CHANGE, payload: {componentId, componentState}});
+}
+
+
+//============================ Utilities ============================
+//============================ Utilities ============================
+//============================ Utilities ============================
+
+export function isDialogVisible(dialogKey) {
+    return get(flux.getState()[DIALOG_OR_COMPONENT_KEY], [DIALOG_KEY,dialogKey,'visible'],false);
+}
+
+export function getDialogOwner(dialogKey) {
+    const d= get(flux.getState()[DIALOG_OR_COMPONENT_KEY], [DIALOG_KEY,dialogKey], {});
+    return (d.visible && d.ownerId) ? d.ownerId : null;
+}
+
+export function getComponentState(componentKey) {
+    return get(flux.getState()[DIALOG_OR_COMPONENT_KEY], [COMPONENT_KEY,componentKey], {});
+}
+
+/*
+   This controller handles small (one level deep) state of various application components,
+   such as dialogs, tab and collapsible panels, which are not a part of field group
+
+   Structure of the store:
+   dialogOrComponent:
+     dialog:
+        [dialogId]: {
+            visible
+            ownerId
+        }
+     component:
+        [componentId]: savedComponentState
+ */
+
+//======================================== Reducer =============================
+//======================================== Reducer =============================
+//======================================== Reducer =============================
+
+
+function reducer(state={[DIALOG_KEY]: {}, [COMPONENT_KEY]: {}}, action={}) {
+    if (!action.payload || !action.type) return state;
+    switch (action.type) {
+
+        case SHOW_DIALOG  :
+            return showDialogChange(state,action);
+
+        case HIDE_DIALOG  :
+            return hideDialogChange(state,action);
+
+        case HIDE_ALL_DIALOGS  :
+            return hideAllDialogsChange(state,action);
+
+        case COMPONENT_STATE_CHANGE  :
+            return changeComponentState(state,action);
+
+        default:
+            return state;
+    }
+}
+
+
+//============ private functions =================================
+//============ private functions =================================
+//============ private functions =================================
+
+const showDialogChange= function(state,action) {
+    var {dialogId,ownerId}= action.payload;
+    const dialogsState = state[DIALOG_KEY];
+
+    if (!dialogsState[dialogId] || !dialogsState[dialogId].visible || ownerId!==dialogsState[dialogId].ownerId) {
+        state = update(state,
+                        {[DIALOG_KEY] :
+                            {$merge: {
+                                [dialogId]: {
+                                    visible: true,
+                                    ownerId
+                                }
+                            }}});
+    }
```

Comment:
For deeper updates, the advantages of update
<https://facebook.github.io/react/docs/update.html> are more prominent.

Writing something like this:

state = update(state,
    {
        [COMPONENT_KEY]: {[componentId]: {$merge: componentState}}
    });

would require 2 clone calls. The deeper you go, the more levels you have to
clone.



2016-04-20 9:06 GMT-07:00 Trey Roby <notifications@github.com>:

> In src/firefly/js/core/ComponentCntlr.js
> <https://github.com/Caltech-IPAC/firefly/pull/63#discussion_r60438198>:
>
> > +//============ private functions =================================
> > +
> > +const showDialogChange= function(state,action) {
> > +    var {dialogId,ownerId}= action.payload;
> > +    const dialogsState = state[DIALOG_KEY];
> > +
> > +    if (!dialogsState[dialogId] || !dialogsState[dialogId].visible || ownerId!==dialogsState[dialogId].ownerId) {
> > +        state = update(state,
> > +                        {[DIALOG_KEY] :
> > +                            {$merge: {
> > +                                [dialogId]: {
> > +                                    visible: true,
> > +                                    ownerId
> > +                                }
> > +                            }}});
> > +    }
>
> An Alternative:
> const clone = (obj,params={}) => Object.assign({},obj,params);
> Once defined:
>
>      state = clone(state,
>                        {[DIALOG_KEY] :
>                            clone(state[DIALOG_KEY],
>                                { [dialogId]: {
>                                    visible: true,
>                                    ownerId
>                                } })});
>
> It is more clear what a function is and what an object is.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Caltech-IPAC/firefly/pull/63/files/8c333937c6238fe436e6f7b0e6b24b6a14165fed#r60438198>
>